### PR TITLE
Refactor Baccano

### PR DESCRIPTION
**Pull request by tabfugnic**
* Build out basic test suite
* Move tests to their own folder
* `rake test` can be used to run test suite
* Add minitest as dependency
* Add `Feature` class to manage individual info
* Create `Fallback` method for cases when standard items do not apply
* Move all libraries into `lib`
* Add lib to load path for more simple require statements
* Created ColorPicker to simplify color picking process
* Created Default to simplify process determining default

### CODE REVIEW

File: lib/avatar.rb [hunk 1]
```diff
@@ -1,16 +1,20 @@
 class Avatar
-  attr_reader :body_color, :hair, :hair_color, :background_color, :skin_tone, :glasses, :hair_background, :eyes
+  attr_reader :body_color, :hair, :background_color, :skin_tone, :glasses, :eyes
 
-
-  def initialize(body_color, background_color, skin_tone, hair, hair_color, glasses, hair_background, eyes)
-    @body_color = body_color
+  def initialize(
```
[the remark this answers is not in the document]
Comment (tabfugnic) [reply lib/avatar.rb:4]: I'm trying! It's hard to keep on whittling it down

File: app.rb [hunk 2]
```diff
@@ -1,7 +1,9 @@
+require "bundler"
 Bundler.require
+$LOAD_PATH.unshift File.expand_path(File.join("..", "lib"), __FILE__)
 require 'yaml'
 require 'digest'
-require_relative 'rng'
+require "rng"
```
Comment (geoffharcourt) [inline app.rb]: Make the ones above double-quotes too so these all look consistent?

**Comment by mike-burns** [comment]
As a high level refactoring: LGTM!

It exposes some things that can be improved (e.g. handle YAML parse errors) but let's get this in first.

File: views/index.erb [hunk 3]
```diff
@@ -5,13 +5,13 @@
...
     <path id="shadow"  fill="<%= @avatar.shadow_color.css_rgba %>" d="M119,45l81,67l0,88l-97.2,0l16.2,-155Z"/>
-    <%= erb @avatar.hair_background rescue nil %>
+    <%= erb @avatar.hair.supplemental_partial, locals: { hair: @avatar.hair } %>
     <path id="body" fill="<%= @avatar.body_color.css_rgba %>" d="M126.7,113.4c-3.8,-5 -15.2,-11.4 -26.7,-11.4c-11.4,0 -22.9,6.4 -26.7,11.4c-12.7,16.7 -21,43.2 -21,65.9c0,3 0.4,20.7 0.4,20.7l94.5,0c0,0 0.4,-17.6 0.4,-20.6c0.1,-22.8 -8.3,-49.4 -20.9,-66Z"/>
     <path id="face" fill="<%= @avatar.skin_tone.css_rgba %>" d="M132.6,74.3c0,12.5-5.7,27-14.4,36.1c-2.6,2.7-10.4,6.2-18.2,6.2s-15.6-3.5-18.2-6.2
        c-8.7-9.1-14.4-23.7-14.4-36.1c0-19.9,14.6-36.1,32.6-36.1S132.6,54.4,132.6,74.3L132.6,74.3z"/>
-    <%= erb @avatar.eyes %>
-    <%= erb @avatar.glasses %>
-    <%= erb @avatar.hair %>
+    <%= erb @avatar.eyes.partial>
```
Comment (geoffharcourt) [inline views/index.erb]: Broken ERB tag.

**Review by geoffharcourt — DISMISSED** [review]
Ack, ran out of time, but found a syntax error.

File: lib/color_picker.rb [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+  def color_contrast
+    @color_contrast ||= generate_contrasting_color
+  end
+
+  def color
+    @color ||= generate_random_color
+  end
+
+  def color_from_file
+    @color_from_file ||= Color::RGB.new(*sample_from_file)
```
Comment (JoelQ) [inline lib/color_picker.rb:20]: Should these memoized functions instead be ivars set in the constructor + a reader? Typically, memoized readers really want just read data that is set in the constructor.

**Comment by JoelQ** [comment]
I like these changes. There are a lot of changes though! Since a lot of the changes are independent, I think this could be broken up into 4-5 smaller commits/PRs that would be easier to review.

**Comment by JoelQ** [comment]
Is there any reason we decided to use minitest instead of RSpec on this project?

**Comment by whmii** [comment]
@JoelQ not sure on the test suite but prior to this PR there was no testing. I'm not sure what changes need to be made but feel free to hack on this. I just don't want the rest of the repo to get so far ahead of this PR that it become irrelevant and eri c s work is wasted.

**Pull request merged**